Refactoring out DaCapo hacks into the DaCapo class

When running DaCapo with callbacks (by overriding the `callback` of an instance of `DaCapo`), it uses `-cp` instead of `-jar` to run DaCapo, which bypasses the MANIFEST file in the jar, causing issues like below

```
java.lang.reflect.InvocationTargetException
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.dacapo.harness.Cassandra.prepare(Cassandra.java:124)
    at org.dacapo.harness.Benchmark.run(Benchmark.java:205)
    at org.dacapo.harness.TestHarness.runBenchmark(TestHarness.java:225)
    at org.dacapo.harness.TestHarness.main(TestHarness.java:170)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at Harness.main(Unknown Source)
Caused by: java.lang.ExceptionInInitializerError
    at org.apache.cassandra.config.DatabaseDescriptor.guessFileStore(DatabaseDescriptor.java:1193)
    at org.apache.cassandra.config.DatabaseDescriptor.applySimpleConfig(DatabaseDescriptor.java:550)
    at org.apache.cassandra.config.DatabaseDescriptor.applyAll(DatabaseDescriptor.java:350)
    at org.apache.cassandra.config.DatabaseDescriptor.daemonInitialization(DatabaseDescriptor.java:178)
    at org.apache.cassandra.config.DatabaseDescriptor.daemonInitialization(DatabaseDescriptor.java:162)
    at org.apache.cassandra.service.CassandraDaemon.applyConfig(CassandraDaemon.java:816)
    at org.apache.cassandra.service.EmbeddedCassandraService.start(EmbeddedCassandraService.java:49)
    ... 13 more
Caused by: java.lang.RuntimeException: java.lang.IllegalAccessException: access to public member failed: jdk.internal.ref.Cleaner.clean[Ljava.lang.Object;@78fbff54/invokeVirtual, from org.apache.cassandra.io.util.FileUtils (unnamed module @158da8e)
    at org.apache.cassandra.io.util.FileUtils.<clinit>(FileUtils.java:107)
    ... 20 more
Caused by: java.lang.IllegalAccessException: access to public member failed: jdk.internal.ref.Cleaner.clean[Ljava.lang.Object;@78fbff54/invokeVirtual, from org.apache.cassandra.io.util.FileUtils (unnamed module @158da8e)
    at java.base/java.lang.invoke.MemberName.makeAccessException(MemberName.java:942)
    at java.base/java.lang.invoke.MethodHandles$Lookup.checkAccess(MethodHandles.java:2201)
    at java.base/java.lang.invoke.MethodHandles$Lookup.checkMethod(MethodHandles.java:2141)
    at java.base/java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:2285)
    at java.base/java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(MethodHandles.java:2278)
    at java.base/java.lang.invoke.MethodHandles$Lookup.unreflect(MethodHandles.java:1747)
    at org.apache.cassandra.io.util.FileUtils.<clinit>(FileUtils.java:96)
```

The current workaround changes `JavaBenchmark` instead of `DaCapo`, which can be confusing for people using other Java benchmark suites. This DaCapo specific hack should be moved to `DaCapo`

https://github.com/anupli/running-ng/blob/4a8a9a027aaf93acc03b3960eb3b9eb470dd222c/src/running/benchmark.py#L208

This DaCapo specific hack should be moved to `DaCapo`, and be conditional on whether `-jar` is used.

https://github.com/anupli/running-ng/blob/4a8a9a027aaf93acc03b3960eb3b9eb470dd222c/src/running/suite.py#L182

Code comment and documentation need to be added.

--- a/src/running/benchmark.py
+++ b/src/running/benchmark.py
@@ -1,10 +1,9 @@
-import errno
 import logging
 import subprocess
 import sys
 from time import sleep
-from typing import Any, List, Optional, Tuple, Union, Dict
-from running.runtime import D8, JavaScriptCore, OpenJDK, Runtime, DummyRuntime, SpiderMonkey
+from typing import Callable, Sequence, TypeVar, List, Optional, Tuple, Union, Dict
+from running.runtime import D8, JavaScriptCore, Runtime, DummyRuntime, SpiderMonkey
 from running.modifier import *
 from running.util import smart_quote, split_quoted
 from pathlib import Path
@@ -12,7 +11,6 @@ from copy import deepcopy
 from running import suite
 import os
 from enum import Enum
-import pty
 
 COMPANION_WAIT_START = 2.0
 
@@ -24,8 +22,11 @@ class SubprocessrExit(Enum):
     Dryrun = 4
 
 
+B = TypeVar('B', bound='Benchmark')
+
+
 class Benchmark(object):
-    def __init__(self, suite_name: str, name: str, wrapper: Optional[str] = None, timeout: Optional[int] = None, override_cwd: Optional[Path] = None, companion: Optional[str] = None, **kwargs):
+    def __init__(self, suite_name: str, name: str, wrapper: Optional[str] = None, timeout: Optional[int] = None, override_cwd: Optional[Path] = None, companion: Optional[str] = None, runtime_specific_modifiers_strategy: Optional[Callable[[Runtime], Sequence[Modifier]]] = None, **kwargs):
         self.name = name
         self.suite_name = suite_name
         self.env_args: Dict[str, str]
@@ -43,6 +44,12 @@ class Benchmark(object):
         # ignore the current working directory provided by commands like runbms or minheap
         # certain benchmarks expect to be invoked from certain directories
         self.override_cwd = override_cwd
+        self.runtime_specific_modifiers_strategy: Callable[[
+            Runtime], Sequence[Modifier]]
+        if runtime_specific_modifiers_strategy is not None:
+            self.runtime_specific_modifiers_strategy = runtime_specific_modifiers_strategy
+        else:
+            self.runtime_specific_modifiers_strategy = lambda _runtime: []
 
     def get_env_str(self) -> str:
         return " ".join([
@@ -55,7 +62,10 @@ class Benchmark(object):
         # also to type check https://mypy.readthedocs.io/en/stable/common_issues.html#variance
         return list(self.wrapper)
 
-    def attach_modifiers(self, modifiers: List[Modifier]) -> Any:
+    def get_runtime_specific_modifiers(self, runtime: Runtime) -> Sequence[Modifier]:
+        return self.runtime_specific_modifiers_strategy(runtime)
+
+    def attach_modifiers(self: B, modifiers: Sequence[Modifier]) -> B:
         b = deepcopy(self)
         for m in modifiers:
             if self.suite_name in m.excludes:
@@ -144,7 +154,7 @@ class BinaryBenchmark(Benchmark):
     def __str__(self) -> str:
         return self.to_string(DummyRuntime(""))
 
-    def attach_modifiers(self, modifiers: List[Modifier]) -> 'BinaryBenchmark':
+    def attach_modifiers(self, modifiers: Sequence[Modifier]) -> 'BinaryBenchmark':
         bb = super().attach_modifiers(modifiers)
         for m in modifiers:
             if self.suite_name in m.excludes:
@@ -182,7 +192,7 @@ class JavaBenchmark(Benchmark):
     def __str__(self) -> str:
         return self.to_string(DummyRuntime("java"))
 
-    def attach_modifiers(self, modifiers: List[Modifier]) -> 'JavaBenchmark':
+    def attach_modifiers(self, modifiers: Sequence[Modifier]) -> 'JavaBenchmark':
         jb = super().attach_modifiers(modifiers)
         for m in modifiers:
             if self.suite_name in m.excludes:
@@ -205,12 +215,6 @@ class JavaBenchmark(Benchmark):
         cmd = super().get_full_args(runtime)
         cmd.append(runtime.get_executable())
         cmd.extend(self.jvm_args)
-        if isinstance(runtime, OpenJDK):
-            if runtime.release >= 9:
-                cmd.extend([
-                    "--add-exports",
-                    "java.base/jdk.internal.ref=ALL-UNNAMED"
-                ])
         cmd.extend(self.get_classpath_args())
         cmd.extend(self.program_args)
         return cmd
@@ -226,7 +230,7 @@ class JavaScriptBenchmark(Benchmark):
     def __str__(self) -> str:
         return self.to_string(DummyRuntime("js"))
 
-    def attach_modifiers(self, modifiers: List[Modifier]) -> 'JavaScriptBenchmark':
+    def attach_modifiers(self, modifiers: Sequence[Modifier]) -> 'JavaScriptBenchmark':
         jb = super().attach_modifiers(modifiers)
         for m in modifiers:
             if self.suite_name in m.excludes:

--- a/src/running/command/minheap.py
+++ b/src/running/command/minheap.py
@@ -108,6 +108,8 @@ def run_with_persistence(result: Dict[str, Any], minheap_dir: Path, result_file:
                     continue
                 print("\t {}-{} ".format(b.suite_name, b.name), end="")
                 mod_b = b.attach_modifiers(mods)
+                mod_b = b.attach_modifiers(
+                    b.get_runtime_specific_modifiers(runtime))
                 minheap = minheap_one_bm(
                     suite, runtime, mod_b, maxheap, minheap_dir, attempts)
                 print("minheap {}".format(minheap))

--- a/src/running/command/runbms.py
+++ b/src/running/command/runbms.py
@@ -114,6 +114,7 @@ def get_hfacs(heap_range: int, spread_factor: int, N: int, ns: List[int]) -> Lis
 def run_benchmark_with_config(c: str, b: Benchmark, runbms_dir: Path, size: Optional[int], fd: Optional[BinaryIO]) -> Tuple[bytes, SubprocessrExit]:
     runtime, mods = parse_config_str(configuration, c)
     mod_b = b.attach_modifiers(mods)
+    mod_b = b.attach_modifiers(b.get_runtime_specific_modifiers(runtime))
     if size is not None:
         mod_b = mod_b.attach_modifiers([runtime.get_heapsize_modifier(size)])
     if fd:

--- a/src/running/plugin/runbms/__init__.py
+++ b/src/running/plugin/runbms/__init__.py
@@ -1,3 +1,4 @@
+from running.plugin.runbms.copyfile import CopyFile
 from pathlib import Path
 from typing import Any, Dict, Optional, TYPE_CHECKING
 if TYPE_CHECKING:
@@ -59,7 +60,6 @@ class RunbmsPlugin(object):
 
 # !!! Do NOT remove this import nor change its position
 # This is to make sure that the plugin classes are correctly registered
-from running.plugin.runbms.copyfile import CopyFile
 if TYPE_CHECKING:
     from running.plugin.runbms.zulip import Zulip
 else:

--- a/src/running/suite.py
+++ b/src/running/suite.py
@@ -1,6 +1,8 @@
 from pathlib import Path
-from typing import Any, Dict, Optional, Union
+from typing import Any, Dict, Optional, Union, List, Sequence
 from running.benchmark import JavaBenchmark, BinaryBenchmark, Benchmark, JavaScriptBenchmark
+from running.runtime import OpenJDK, Runtime
+from running.modifier import JVMArg, Modifier
 import logging
 from running.util import register, split_quoted
 
@@ -201,6 +203,17 @@ class DaCapo(JavaBenchmarkSuite):
             program_args.extend(["-s", size])
         # Name of the benchmark
         program_args.append(bm_name)
+
+        def strategy(runtime: Runtime) -> Sequence[Modifier]:
+            modifiers = []
+            if isinstance(runtime, OpenJDK):
+                if runtime.release >= 9:
+                    modifiers.append(JVMArg(
+                        name="add_exports",
+                        val="--add-exports java.base/jdk.internal.ref=ALL-UNNAMED"
+                    ))
+            return modifiers
+        
         return JavaBenchmark(
             jvm_args=[],
             program_args=program_args,
@@ -209,7 +222,8 @@ class DaCapo(JavaBenchmarkSuite):
             companion=self.get_companion(bm_name),
             suite_name=self.name,
             name=name,
-            timeout=timeout
+            timeout=timeout,
+            runtime_specific_modifiers_strategy=strategy
         )
 
     def get_minheap(self, bm: Benchmark) -> int:

--- a/tests/test_suite.py
+++ b/tests/test_suite.py
@@ -77,3 +77,49 @@ def test_dacapo_timing_iteration():
     fop_converge = c.get("benchmarks")["dacapo2006"][1]
     assert "-n 3" in fop.to_string(DummyRuntime("java"))
     assert "-converge" in fop_converge.to_string(DummyRuntime("java"))
+
+
+def test_dacapo_openjdk_9_workaround():
+    c = Configuration({
+        "suites": {
+            "dacapo2006": {
+                "type": "DaCapo",
+                "release": "2006",
+                "path": "/usr/share/benchmarks/dacapo/dacapo-2006-10-MR2.jar",
+                "timing_iteration": 3
+            }
+        },
+        "benchmarks": {
+            "dacapo2006": [
+                "fop"
+            ]
+        },
+        "runtimes": {
+            "jdk8": {
+                "type": "OpenJDK",
+                "release": 8,
+                "home": "/usr/lib/jvm/temurin-8-jdk-amd64"
+            },
+            "jdk11": {
+                "type": "OpenJDK",
+                "release": 11,
+                "home": "/usr/lib/jvm/temurin-11-jdk-amd64"
+            },
+        },
+        "configs": [
+            "jdk8",
+            "jdk11"
+        ]
+    })
+    c.resolve_class()
+    from running.benchmark import JavaBenchmark
+    fop: JavaBenchmark
+    fop = c.get("benchmarks")["dacapo2006"][0]
+    jdk8 = c.get("runtimes")["jdk8"]
+    jdk11 = c.get("runtimes")["jdk11"]
+    fop_jdk8 = fop.attach_modifiers(fop.get_runtime_specific_modifiers(jdk8))
+    fop_jdk11 = fop.attach_modifiers(fop.get_runtime_specific_modifiers(jdk11))
+    print(fop_jdk8.to_string(jdk8))
+    print(fop_jdk11.to_string(jdk11))
+    assert "add-exports" not in fop_jdk8.to_string(jdk8)
+    assert "add-exports" in fop_jdk11.to_string(jdk11)

